data/persons.txt

```diff
@@ -897,7 +897,7 @@ person "Prototype B3-CC4"
 
 	phrase # The ones below are EVN Krypt Pod references
 		word
-			"<"
+			"&lt;"
```

Comment:
It causes no actual problem with the existing contents because this PR uses a text as a plain text if it contains an invalid markup. This change is only to avoid the fallback in run-time.

This PR introduces an incompatibility when a ship's name contains "&amp;amp;", "&amp;lt;", "&amp;gt;", or a valid markup sequence. I would like to provide the players with more capabilities, but we can escape/recover the sequence when loading/saving it.

![Peek 2020-12-30 20-41](https://user-images.githubusercontent.com/22392249/103350737-42e16a00-4ae4-11eb-87b4-060320a4e78c.gif)
I input '>', and it becomes a valid markup, then '<' causes it's an invalid.

We can remove the pango markup from this PR, but I have used it in my Japanese translation. Many names of the ship models and the outfits are truncated if I don't use a condensed font.

What do you think?

Reply:
We should probably escape all markup found in strings that get provided to Font, so that the user would always see e.g. `<span color="red">red one</span>`. Then we would not need to worry about users transmitting malicious markup via save files or plugins, or the entered text undergoing visual transformations while typing.

For other languages we could provide "system-level" markup as a part of the `drawingSettings`, `Layout`, etc. Then one would be able to apply the condensed font or other needed changes to improve the display, and know which markup is safe vs which markup needs to be escaped.

Reply:
I agree that the input fields should be completely escaped.

For a malicious markup, a pango markup makes easy to create a huge texture. I thought Endless Sky didn't care any problems that a plugin causes, but we can limit the size of a texture if we want. For example, 10000x10000.

An implementation of "system-level" markup is to filter out any unacceptable markups. If we would adopt this implementation, I need to know what pango markups are unacceptable. What pango markups should be filtered out?

By the way, the current implementation of this PR filters out any character references except "&amp;amp;", "&amp;lt;", and "&amp;gt;". It's to minimize the incompatible of the ship names, so we don't need this filter if any names' of ships and <first> <last> are completely escaped. (<first> <last> can contains '&'. I notice it now.)

I suggest we adopt one follows:
1. Escape any texts in the input fields and any names of the ships and <first> <last>, and recover the names when saving for compatibility. (For example, the ship's name "&amp;amp;" in the save file becomes "&amp;amp;amp;" when loading, then recovers to "&amp;amp;" when saving.)
2. "1." + Introduce a restriction against pango markups, that is to create "system-level" markups.
3. Remove all markups and character references (for now or permanently).

Easy to implement: 3 > 1 > 2
Expressiveness: 1 > 2 > 3

I have experience of using attributes 'font', 'size', 'rise', and 'letter_spacing' for Japanese translation. If I have a free to choose/create a font, I need only 'font'. (I have used 'size', 'rise', and 'letter_spacing' for translations of Wesnoth UMC.) To create a font using Private Use Area avoids to select fonts, but it needs to create a translation file hardly to maintain. I guess those who would create a font are not many, though those who care about layout seem to be not many. Disclaimer: I don't know the text layout except for the Japanese.

Reply:
My thought for "system-level" markup was that inline, user-sourced string markup would be disallowed, but markup associated with the language & font could be "wrapped around" the provided text and then provided to pango & cairo for rendering. So this would disallow bold / italic / coloring on individual words (e.g. those obtained via `Phrase::Get()`, outfit/mission descriptions, government names, etc.) but still allow the translator to choose size, rise, etc.


Is that flexible enough? Or do we need to support certain "word-level" pango markup attributes, too, such that we can't disallow inline markup on user-provided strings? (I have no experience in translation 😅) 
If we need some word-level attributes, probably it is much simpler to allow all markup at first and only later restrict the allowed inline user markup if there are reports of issues.

My stance on plugins and 3rd party content is that we shouldn't make it _trivial_ to create a disruptive experience / unplayable game / system crash. If you have ideas about how we can achieve that, please share 👍

Reply:
The translation with the gettext basically only replaces a string to another string, so a translator cannot do what a content's creator cannot do. We may consider only what a content's creator should be allowed. A translation was only an example that I could provide.

If a markup from a content's creator should be disallowed, I suggest removing all markups and character references. It's easy to implement and to understand. If someone needs a markup, they will be able to post a PR.

However, what a pango markup causes are only a lack of the memory and a disturbance in a display as far as I know. We can restrict the maximum size of a texture to prevent a lack of memory. For a disturbance in a display, we should improve the UI code. Cf. #972, #1768, #2022. He has always impressed me, seriously.

Reply:
Let's go with limited texture size & option 1 above. We have other filed issues for the work to improve the UI when text overflows the game's expected bounds, so we don't need to handle that in this PR.